AWS Patches for Libfabric Host Proxy Transport

a-szegel opened the pull request:
Hi Nvidia!  Thanks for putting NVSHMEM on Github!

a-szegel committed:
transport/libfabric: Make gdr put-signal send 32 byte signal packet

The EFA Provider has a 32 byte inline send data size. Previously, the put-signal
implementation for gdrcopy emulated atomics called the already existing atomic
functions in order to optimize code re-use. The size of an atomic operation wire
data packet is 72 bytes, which is larger than the EFA provider's 32 byte inline
send size. This change modifies the wire data for the signal part of the put-signal
operation to be 32 bytes, at the expense of code reuse.  When the signal wire data
packet is recived by the RX side, the RX side will re-format that data in-place to
re-use the existing emulated gdrcopy atomic functions.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

a-szegel committed:
transport/libfabric: Fix Bug In Error Logging Logic

A logging bug exists where the user passes status into try_again() and
then passes status into fi_strerror() to get the error code number to
string meaning.  The try_again() function changes the status to
NVSHMEMX_ERROR_INTERNAL = 7, which results in the wrong libfabric error
code being printed, and confuses the debugger.  This commit also adds a
warning message to try_again to print the real error code from
libfabric.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

a-szegel committed:
transport/libfabric: Fix efa put-signal 0 byte write bug

EFA's put-signal implementation uses an RX side hashmap to hold
the signal update until all corresponding write's have been
completed. This implementation did not handle the edge case of
0 writes associated with a put signal operation. A put-signal
operation with 0 writes is just a signal operation, and should
immediately set the signal value.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

a-szegel committed:
transport/libfabric: Move put-signal map/counters to ep

Before this patch, the libfabric host proxy transport had 1 map, and
1 counter for both EP's (proxy EP, and host EP).  This can lead to
potential threading issues if both EP's are getting progressed at the
same time.  The way that NVSHMEM transports deal with this potential
race condition is to seperate the resources for host ops, and proxy ops,
and never miss them.  This patch splits the map, and the put-signal
sequence counters by host and proxy (1 copy for each).  This allows us
to not use mutex locks, because these resources will only ever be
accessed by one thread.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

a-szegel committed:
transport/libfabric: Switch efa provider to efa-direct

The EFA provider has two different “modes” (not to be confused with mode bits),
efa and efa-direct. Each of the "modes" are seperate providers that come with
their own libfabric resources (info, fabric, domain, ep, cq, and
counter). The efa “mode” uses software protocols in order to provide libfabric
api completeness. The efa-direct “mode” has no software protocols, and directly
exposes the capabilities of the device to the user.

In order to use the "efa-direct" mode, the application must adhere to
the following modes:
  FI_MR_LOCAL: Datapath can only use pre-registered buffers
  FI_CONTEXT2: App provides 64 bytes along with every message for the
               provider to use, App must not touch this memory

This change does the following actions to support efa-direct:

1. Add a fi_context2 struct to nvshmemt_libfabric_gdr_op_ctx, and pass
   it into every libfabric datapath api call
2. Unify the send/recv bufferpools into 1 big buffer, and register that
   buffer with the provider. Always pass in the MR to all data path api
   calls
3. Modify the hints structure in the info call to ensure that we always
   get efa-direct provider
4. Specify CQ size of 32k to avoid CQ overflow

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

a-szegel committed:
transport/libfabric: Make efa put-signal seq_cnt 32 bits

The libfabric API fi_writedata() takes a 64 bit immediate data
parameter, but the EFA provider only supports 32 bits of immediate CQ
data. Switch the sequence counter to use 32 bits to match EFA's write
with immediate capability.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

a-szegel committed:
transport/libfabric: Make put-signal seq_cntr 28 bits

This commit makes the put-signal sequence counter 28 bits from 32 bits.
Previously, all NVSHMEM write with immediate operations were assumed to
be for put signal, but now we have the freedom to use write w/imm for
15 other non-put signal use cases as well.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

a-szegel committed:
transport/libfabric: Make fence() correct with staged_atomics

The libfabric transport's fence() implementation for staged_atomics
depended on completion ordering (comp_order) which was not requested,
and has been deprecated from the libfabric API. The fence()/quiet()
implementation would block until all TX completions were received for
every submitted operation, but that does not guarantee that the atomic
has been applied at the peer. For the following program:

atomic(5)
fence()
atomic(8)

There is a race condition that existed where the fence finishes but the
first atomic has not been processed by the RX side.  The TX side sends
the second atomic, and both of them are sitting in the devices RX
completion queue at the same time. The libfabric API makes no guarantees
on completion ordering so atomic(8) could happen before atomic(5).

This patch fixes this race by adding a 0 byte write w/imm ack for every
staged_atomic operation. When the TX side submits an atomic, they
increase the submitted ops by 2, and the fence()/quiet() will not
complete until the ack comes back from the reciver saying that the
atomic was applied.

This patch also adds a 4 bit header to the immediate data allowing the
immediate data to be used for more than just put_signal_unordered.

For a 0 byte write w/imm, the efa device currently requires valid
paramters for local buffer, local buffer desc, rptr, and rkey even
though no data is being transmitted.  In order to meet these
requirements, a registered write target variable was added to libfabric
state, and its address/rkey are exchanged during bootstrap.  I attempted
to re-use the same atomic that had the operation performed on it, but
this was not possible because the nvshmemi_state global is not exposed
to the transport libraries.

This change is expected to make fence()/quiet() calls more expensive for
libfabric providers that use staged_atomics, but is required for
correctness.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

a-szegel committed:
transport/libfabric: Correctly size staged_atomic buffer pools

tx_attr->size represents the maximum number of posted TX operations on
the NIC at any one time.  rx_attr->size represents the maximum number of
posted RX buffers on the NIC at any one time.  There is no need to
multiply fixed NIC constant values by the number of PE's b/c the extra
buffers will just be wasted space.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

a-szegel committed:
transport/libfabric: Get staged_amo src_addr from fi_cq_readfrom()

This change modifies the contents of the staged_amo message to send the
PE instead of fi_addr_t of the PE.  The fi_addr_t of the sender is
available from fi_cq_readfrom() on the RX side, so this information is
no longer required to be sent over the wire. This change switches that
field within the message structure to PE to prep for EFA GDA re-use of
the libfabric host proxy transport as a nested transport.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

a-szegel committed:
transport/libfabric: Only 1 put-signal seq_cntr per EP

Previously, there was 1 put-signal seq_cntr per EP per Peer. This was
done to make the rate of the sequence counters go up slower to prevent
possible overflow collision race conditions.  With 28 bits, the max
value of the sequence counter is 268 million, and with a packet rate of
~2 million packets per second, the sequence counter is not likely to
overflow anyways.  This change reduces the number of sequence counters
to be 1 sequence counter per EP (proxy or host thread) which should be
nicer on cache.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

rauteric committed:
perftest: fix uninitialized variable in table print

The test did not initialize `tokens` array, so if `subjob_name` did not
contain any tokens, the entries of `tokens` were left uninitialized,
leading to a segfault.

Add initialization of `tokens` entries and overflow checking, and remove
leak of tokens array.

Signed-off-by: Eric Raut <eraut@amazon.com>

rauteric committed:
transport/libfabric: Track put-signal sequence numbers

We are using an index to match put-with-signal payload and signal data.
Although highly unlikely, it was possible for the index to wrap around
and conflict with a previous incomplete put-with-signal operation.

This patch handles overflow explicitly. A category bit(s) is reserved in
the index number. Before re-using a category, the code make sure all
acks for the category have completed from the previous round. As long as
there is more than one category, blocking should be extremely unlikely.

Signed-off-by: Eric Raut <eraut@amazon.com>

prajwal1210 commented:
When trying to compile these changes, I get:

```
/usr/local/nvshmem_src/src/modules/transport/ucx/ucx.cpp: In function 'int nvshmemt_ucx_quiet(nvshmem_transport*, int, int)':
/usr/local/nvshmem_src/src/modules/transport/ucx/ucx.cpp:1035:38: error: too few arguments to function 'int nvshmemt_ucx_progress(nvshmem_transport_t, int)'
 1035 |                 nvshmemt_ucx_progress(tcurr);
      |                 ~~~~~~~~~~~~~~~~~~~~~^~~~~~~
/usr/local/nvshmem_src/src/modules/transport/ucx/ucx.cpp:55:5: note: declared here
   55 | int nvshmemt_ucx_progress(nvshmem_transport_t transport, int is_proxy);
      |     ^~~~~~~~~~~~~~~~~~~~~
/usr/local/nvshmem_src/src/modules/transport/ucx/ucx.cpp:1039:38: error: too few arguments to function 'int nvshmemt_ucx_progress(nvshmem_transport_t, int)'
 1039 |                 nvshmemt_ucx_progress(tcurr);
      |                 ~~~~~~~~~~~~~~~~~~~~~^~~~~~~
/usr/local/nvshmem_src/src/modules/transport/ucx/ucx.cpp:55:5: note: declared here
   55 | int nvshmemt_ucx_progress(nvshmem_transport_t transport, int is_proxy);
      |     ^~~~~~~~~~~~~~~~~~~~~
```

Looking at the code, it seems the nvshmemt_ucx_progress is not always called with the proxy argument, and also the declaration and definition of the function in the ucx.cpp file differ.

a-szegel commented:
Hey @prajwal1210, Thanks for catching this bug.  I fixed it, and tested that UCX provider compiles now. Thanks!

a-szegel committed:
transport/libfabric: Protect put-signal signal update with mutex

Previously, the put-signal completion logic would set the signal value
without grabbing the gdrRecvMutex lock. The gdrRecvMutex lock is
requried to make sure that host submited atomic ops and gpu submitted
atomic ops implemented via GDR copy are serialized with each other.
The user could hit this bug if they submitted put_signal operations from
the host thread and the gpu at the same time, targeting the same PE,
using the same signal object, which would cause atomicity issues.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

a-szegel committed:
transport/libfabric: Properly propagate error codes on completion path

This patch checks return values of functions, and enforces that failures
in the completion handling path will cause application abort.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

prajwal1210 commented:
Hi @a-szegel!

Thanks for the detailed PR. I actually understood a lot about the libfabric backend based on your commit messages. I am working with NVSHMEM over Slingshot and was curious about this commit message:

```
    The libfabric transport currently supports 3 providers (CXI, EFA,
    VERBS). The CXI/Verbs provider both support FI_ORDER_WAW, and have NIC's
    that capable of supporting atomic operations. The EFA provider does not
    support FI_ORDER_WAW, and the NIC is not capable of hardware atomics.
    This commit optimizes the EFA Providers implementation of
    put_with_signal, and future commits can further optimize cxi and verbs
    provider by removing the fence and requesting FI_ORDER_WAW.
 ```

I wanted to work on this but is FI_ORDER_WAW enough for CXI/Verbs? I believe that adding this to the tx_attr and rx_attr ensures the rma and atomic ops are sent and received in order but they might not be completed in order, correct?

One would need to use a completion order as well here, correct?

a-szegel commented:
Hi @prajwal1210,

I am not 100% confident on this answer, so please double check me with Libfabric/IB spec.

> FI_ORDER_WAW
Write after write. If set, RMA and atomic write operations are transmitted in the order submitted relative to other RMA and atomic write operations. If not set, RMA and atomic writes may be transmitted out of order from their submission.

I believe that FI_ORDER_WAW is enough to guarantee the atomics get executed in order at the target, which should be enough for you to set fence=NULL.  

Libfabric 2.0 deprecated completion ordering.  Completion Ordering refers to the order in which completions are written into the completion queue.  Slingshot sets FI_SELECTIVE_COMPLETION, which means that they are not generating completions for successful transfers, so comp_order, even if supported would not be helpful to you.

seth-howell commented:
That depends if the underlying network guarantees in order submission/delivery of operations posted to the queue. I know that's the case for IB, I am not sure for slingshot.

prajwal1210 commented:
So, from what I understand, Slingshot does offer guarantees for the submission/delivery of rma and atomic writes when using FI_ORDER_WAW on both tx and rx attributes. The only bit I am confused about is whether it guarantees that the received operations will be put in the completion queue in order. I can dig a little bit more to find out. I believe it does based on @a-szegel's comment.  

Are there any tests one can run when contributing to NVSHMEM to ensure the correctness of the changes implemented?

a-szegel commented:
Hey @prajwal1210,

As soon as you set FENCE=NULL, completions don't matter anymore with respect to ordering, so it doesn't matter what completion parameters you set. FENCE=NULL means that the network is fully ordered, and operations will be executed at the target in the order that they were submitted with respect to a single PE.

This is where fence is [invoked](https://github.com/NVIDIA/nvshmem/blob/devel/src/host/proxy/proxy.cpp#L853) from the proxy code.

For example: 
```
put(dest_pe=1, dest_addr=1, val=5);
fence();   // this is always no-op when tcurr->host_ops.fence=NULL
put(dest_pe=1, dest_addr=1, val=7, );
```
The transport must guarantee that the contents of dest_addr is always 7 for the program to be considered correct.

If you want to keep discussing this, can you please open an issue and I will happily engage with you there.  I would like to this discussion thread focused on the commits within the PR if possible.

Thanks!

a-szegel committed:
transport/libfabric: Always request FI_DELIVERY_COMPLETE

Always request FI_DELIVERY_COMPLETE for correctness of nvshmem_quiet()
because RDMA operations must complete on the RX side before generating
TX completions.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

a-szegel committed:
Fix NVSHMEM_DEVEL build

Fix warnings on ubuntu2404, gcc13.3 to make the devel build succeed.

Signed-off-by: Seth Zegelstein <szegel@amazon.com>

seth-howell commented:
This has been merged to devel and main and is now available in the repo.
